dali/pipeline/executor/executor.cc

```diff
@@ -356,10 +356,16 @@ int Executor<WorkspacePolicy, QueuePolicy>::InferBatchSize(
     DALI_ENFORCE(bsps[0]->NextBatchSize() == bsps[i]->NextBatchSize(),
                  "Batch size must be uniform across an iteration");
   }
-  auto batch_size = bsps[0]->NextBatchSize();
-  assert(batch_size > 0);
-  for (auto &bsp : bsps) {
-    bsp->Advance();
+  try {
+    auto batch_size = bsps[0]->NextBatchSize();
+    assert(batch_size > 0);
+    for (auto &bsp : bsps) {
+      bsp->Advance();
+    }
+  } catch (const std::exception &e) {
```

Comment:
I think it's dirty to catch a generic exception and assume its meaning. There should be a better error resolution upstream (preferred) or better inspection here (passable, but still ugly). The way it is, we can produce truly ridiculous errors here, like:
```
"Failed to acquire the next batch. Make sure, that DALI Pipeline is fed with sufficient amount of data. Bad alloc."
```

Reply:
How about adding specialized exception to `batch_size_provider.h`? `NextBatchSize` and `Advance` could throw this one then
```c++
struct cache_empty : std::out_of_range {};
```
?

Reply:
I don't think that's necessary. However, we already throw `out_of_range` and we could catch that. This at least eliminates everything else we could possibly catch here.